Add `redirect_back` and deprecate `redirect_to :back`

derekprior opened the pull request:
`redirect_to :back` is a somewhat common pattern in Rails apps, but it
is not completely safe. There are a number of circumstances where HTTP
referrer information is not available on the request. This happens often
with bot traffic and occasionally to user traffic depending on browser
security settings.

When there is no referrer available on the request, `redirect_to :back`
will raise `ActionController::RedirectBackError`, usually resulting in
an application error.

`redirect_to_back_or_default` is a helper I have added to many rails
applications to avoid this situation and now I'm proposing it be added
to Rails proper and encouraged in place of `redirect_to :back`.

simi commented:
I think it is common pattern to use `rescue_from` in those situations.

```ruby
class ApplicationController < ActionController::Base
  rescue_from ActionController::RedirectBackError, with: :redirect_to_default

  protected
  def redirect_to_default
    redirect_to root_url
  end
end
```

claudiob commented:
Hello and thanks for your contribution!

> This happens often with bot traffic and occasionally to user traffic depending on browser
security settings.

You are making the point that `redirects_to :back` is not something you can rely on, and I agree with that.
For that reason, I have stopped using `redirects_to :back` altogether a while ago.

Could you provide a real case (with or without code) in which you would still need to redirect to the previous page if the referrer is present while, at the same time, redirect to a default page if not present?

That would be helpful to better understand the value of the PR. Thanks! :ribbon:

derekprior commented:
Thanks for the great questions @claudiob. 

> For that reason, I have stopped using redirects_to :back altogether a while ago.

I largely concur, but as a consultant exposed to dozens of applications, I continue to see `redirect_to :back` in app after app. In some cases it's simple enough to refactor to a static redirect, but in other cases `:back` with a sensible default is really the simplest thing that could possibly (very nearly always :grin:) work.

Consider, for example, an application that manages products. Products can be viewed in many different contexts and from each of these views you are able to "Activate" of "Deactivate" the product. The controller for activations may look like this:

```ruby
class ProductActivationsController < ActionController::Base
  def create
    Product.find(params[:product_id]).activate
    redirect_to :back, notice: "Product activated"
  end

  def destroy
    Product.find(params[:product_id]).deactivate
    redirect_to :back, notice: "Product deactivated"
  end
end
```

Because these controller actions can be called from various locations -- say from the admin products index, the admin product show, the store ui if you are an admin, or any number of various other places, the notion of "from wherever you came" is valuable. You can attempt to pass the return path explicitly as a parameter of the `create` and `destroy` action but then to avoid [unvalidated redirects](https://www.owasp.org/index.php/Unvalidated_Redirects_and_Forwards_Cheat_Sheet). It's a decent amount of code for a case when `:back` works well in all but a handful of cases.

If I were designing this solution in a clean room, I'd deprecate `redirect_to :back` and either rely on a gem for a smarter replacement or replace it with `redirect_back_or_default` so you would never get an exception.

Either way, I continue to see application that have dozens of not so straight-forward to solve `ActionController::RedirectBackError`s in their error trackers. The simplest fix is generally to provide a default rather than the shotgun surgery that would be required to get the correct redirect for all occurrences. I'm not generally thrilled with this as a solution, but from a cost/benefit standpoint it tends to make a lot of sense.

> I think it is common pattern to use rescue_from in those situations.

This can get out of hand if you have different defaults per action. It also relies on exceptions for flow control, which I'd prefer not to do.

derekprior commented:
All that said, I'd totally understand not accepting this PR, but then I'd suggest deprecating `redirect_to :back`.

sgrif commented:
I definitely agree that this should replace `redirect_to :back` with this. Our general philosophy has been to avoid constructs which could allow bad user input to make the server 500 trivially. `rescue_from` is definitely not sufficient. 

That said, I'm not huge on the method name, it feels a bit long and unweildly to me.

derekprior commented:
I agree. If deprecating `redirect_to :back` is on the table then I'd suggest something like `def redirect_back(default:, *args)`. But if the previous method isn't going to be deprecated then I think they are too near eachother.

bogdan commented:
Agree on deprecation of `redirect_to :back`. To the reasons mentioned above, I can add that passing magic symbols to a method seems bad API to me. rescue `ActionController::RedirectBackError` is bad idea when default differs per action.

sgrif commented:
I like `redirect_back(default:, *args)`. I think deprecating `redirect_to :back` is a separate discussion (which I am in favor of), but we can use that method name regardless. My only other concern is that `:default` doesn't necessarily imply anything if I'm reading code and not familiar with this exact problem.

simi commented:
What about `redirect_back(fallback: *args)` :question:

sgrif commented:
`:fallback` doesn't give much more context than `:default`. Maybe `redirect_back(url_if_unknown:)`, though that's a bit verbose (maybe this needs to be verbose?)

simi commented:
I was about to avoid `url` or `path` as part of the argument name.

derekprior commented:
It's not always a URL. You could pass anything that worked with `redirect_to`. I think `fallback` is the best suggestion I've heard so far:

```ruby
def create
  product = Product.find(params[:product_id]
  product.activate
  redirect_back(fallback: product, notice: "Activated")
end
```

In the context of `redirect_back` I think fallback is pretty clear. We could also use `fallback_redirect`?

sgrif commented:
I think that `:fallback` is only clear if you know how back works, and
understand that it may not be there. I don't think that we can assume that.

On Sun, Dec 6, 2015, 10:31 AM Derek Prior <notifications@github.com> wrote:

> It's not always a URL. You could pass anything that worked with
> redirect_to. I think fallback is the best suggestion I've heard so far:
>
> def create
>   product = Product.find(params[:product_id]
>   product.activate
>   redirect_back(fallback: product, notice: "Activated")end
>
> In the context of redirect_back I think fallback is pretty clear. We
> could also use fallback_redirect?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/22506#issuecomment-162327273>.
>

simi commented:
@sgrif yes it is clear only if you know how `:back` works, but that should be explained in the documentation.

I think introducing of `redirect_back`, deprecation of `redirect_to :back` and leaving note in deprecation message and `ActionController::RedirectBackError` message will be great start for this feature.

kaspth commented:
r? @sgrif

sgrif commented:
I think `redirect_back(fallback:)` is the way to go for now. @derekprior Can you update and rebase?

derekprior committed:
Add `redirect_back` for safer referrer redirects

`redirect_to :back` is a somewhat common pattern in Rails apps, but it
is not completely safe. There are a number of circumstances where HTTP
referrer information is not available on the request. This happens often
with bot traffic and occasionally to user traffic depending on browser
security settings.

When there is no referrer available on the request, `redirect_to :back`
will raise `ActionController::RedirectBackError`, usually resulting in
an application error.

`redirect_back` takes a required `fallback_location` keyword argument
that specifies the redirect when the referrer information is not
available.  This prevents 500 errors caused by
`ActionController::RedirectBackError`.

derekprior commented:
@sgrif updated and rebased.

simi commented:
No deprecation of `redirect_to(:back)` for now?

Anyway I think the documentation of `redirect_to(:back)` should mention potential dangers and link to this new method.

sgrif commented:
Let's just go ahead and deprecate, too

derekprior committed:
Deprecate `redirect_to :back`

Applications that use `redirect_to :back` can be forced to 500 by
clients that do not send the HTTP `Referer` (sic) header.
`redirect_back` requires the user to consider this possibility up front
and avoids this trivially-caused application error.

derekprior commented:
@sgrif Deprecation added.

sgrif commented on actionpack/test/controller/redirect_test.rb:
```diff
@@ -187,7 +191,11 @@ def test_relative_url_redirect_with_status_hash
     @request.env["HTTP_REFERER"] = "http://www.example.com/coming/from"
-    get :redirect_to_back_with_status
+
+    ActiveSupport::Deprecation.silence do
```
These should all be `assert_deprecated`

sgrif commented on actionpack/test/controller/redirect_test.rb:
```diff
@@ -236,18 +244,52 @@ def test_redirect_to_url_with_network_path_reference
     assert_equal "http://www.example.com/coming/from", redirect_to_url
   end
 
+  def test_redirect_to_back_is_deprecated
```
We don't need an additional test here, the previous tests changed to `assert_deprecated` are enough.

simi commented on actionpack/lib/action_controller/metal/redirecting.rb:
```diff
@@ -87,6 +103,12 @@ def _compute_redirect_to_location(request, options) #:nodoc:
+        ActiveSupport::Deprecation.warn(<<-MESSAGE.squish)
+          `redirect_to :back` is deprecated and will be removed from Rails 5.1.
+          Please use `redirect_back(fallback: fallback_location)` where
+          `fallback_location` represents the redirect to use if the request has
```
...represents the **redirect** :question: Maybe **location** or *url** will be better to use here.

derekprior replied:
Yeah, I'll use location. It's not always a URL but rather something that will eventually be a URL. I think location is a better indicator of that.

sgrif replied:
What do you think about making the keyword argument `fallback_location`?

On Wed, Dec 16, 2015 at 8:15 AM Derek Prior <notifications@github.com>
wrote:

> In actionpack/lib/action_controller/metal/redirecting.rb
> <https://github.com/rails/rails/pull/22506#discussion_r47787127>:
>
> > @@ -87,6 +103,12 @@ def _compute_redirect_to_location(request, options) #:nodoc:
> >        when String
> >          request.protocol + request.host_with_port + options
> >        when :back
> > +        ActiveSupport::Deprecation.warn(<<-MESSAGE.squish)
> > +          `redirect_to :back` is deprecated and will be removed from Rails 5.1.
> > +          Please use `redirect_back(fallback: fallback_location)` where
> > +          `fallback_location` represents the redirect to use if the request has
>
> Yeah, I'll use location. It's not always a URL but rather something that
> will eventually be a URL. I think location is a better indicator of that.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/22506/files#r47787127>.
>

simi commented:
thanks all :confetti_ball:

emilebosch commented:
Sorry im a bit sad about this. :( I like rails and i like that it optimized for happiness but this doesnt make me happy.

Nooo. you guys didn't just change `redirect_to :back` and completely worked me up. I'm ok with raises redirect errors, they are super loggable and rescuable. And then i know where this referer less stuff comes from.(ips, agents etcetc) But now i have to update all my code & rails engines etc etc.  Am i missing something or does this really just do nothing except add work?

Why didn't you just made a comprise:

```
redirect_to :back, notice:"Yolo", fallback: '/'
```

And add a rails config that raises if fallback isn't present? But people that know what they are doing can just use whatever they want and put that `config.raise_on_no_fallback = false`

juanantoniodamianv commented:
Hi!
Can I check the value of :back, so that if it is a specific url, I will go back, and if not, go back to root?
